File: web/src/lib/importer.ts
/**
 * Manual CSV import for the `transactions` table. Direct TypeScript port of
 * portfolio/importer.py (repo root) — same validation rules, same
 * all-or-nothing behavior, same dedup approach. See that file's docstring
 * and the README for the conventions this encodes (cash-flow quantity/price
 * convention, single-entry model, etc.).
 */

import { createHash } from "node:crypto";
import { parse } from "csv-parse/sync";
import type { PoolClient } from "pg";
import type { AssetClass, ImportStats, TransactionType } from "./types";

const REQUIRED_COLUMNS = [
  "date",
  "asset_symbol",
  "asset_class",
  "type",
  "quantity",
  "price",
  "fees",
  "currency",
  "source",
] as const;

const VALID_ASSET_CLASSES: readonly AssetClass[] = ["stock", "etf", "cash", "gold", "crypto"];
const VALID_TYPES: readonly TransactionType[] = [
  "buy",
  "sell",
  "deposit",
  "withdrawal",
  "interest",
  "dividend",
];

const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

export class CsvValidationError extends Error {}

interface NormalizedRow {
  date: string;
  assetSymbol: string;
  assetName: string | null;
  assetClass: AssetClass;
  type: TransactionType;
  quantity: number;
  price: number;
  fees: number;
  currency: string;
  source: string;
  notes: string | null;
}

function isValidDate(s: string): boolean {
  if (!DATE_RE.test(s)) return false;
  // Reject e.g. 2024-02-30 — Date would otherwise silently roll it forward
  // to March 1st instead of rejecting it.
  const d = new Date(`${s}T00:00:00Z`);
  return d.toISOString().slice(0, 10) === s;
}

/** Parses the CSV and returns one plain object per data row, keyed by header. */
function parseRows(csvText: string): Record<string, string>[] {
  let rawRows: string[][];
  try {
    rawRows = parse(csvText, { skip_empty_lines: true, trim: false }) as string[][];
  } catch (err) {
    throw new CsvValidationError(`Failed to parse CSV: ${(err as Error).message}`);
  }

  if (rawRows.length === 0) {
    throw new CsvValidationError("CSV is empty");
  }

  const header = rawRows[0];
  const missing = REQUIRED_COLUMNS.filter((c) => !header.includes(c));
  if (missing.length > 0) {
    throw new CsvValidationError(`CSV is missing required column(s): ${JSON.stringify(missing)}`);
  }

  return rawRows.slice(1).map((row) => {
    const record: Record<string, string> = {};
    header.forEach((col, i) => {
      record[col] = row[i] ?? "";
    });
    return record;
  });
}

function validateAndNormalizeRow(
  row: Record<string, string>,
  lineNo: number,
): { normalized: NormalizedRow | null; errors: string[] } {
  const errors: string[] = [];
  const get = (field: string) => (row[field] ?? "").trim();

  const dateStr = get("date");
  if (!isValidDate(dateStr)) {
    errors.push(`line ${lineNo}: invalid date ${JSON.stringify(dateStr)} (expected YYYY-MM-DD)`);
  }

  const assetClass = get("asset_class").toLowerCase() as AssetClass;
  if (!VALID_ASSET_CLASSES.includes(assetClass)) {
    errors.push(
      `line ${lineNo}: invalid asset_class ${JSON.stringify(assetClass)} (expected one of ${JSON.stringify(VALID_ASSET_CLASSES)})`,
    );
  }

  const type = get("type").toLowerCase() as TransactionType;
  if (!VALID_TYPES.includes(type)) {
    errors.push(
      `line ${lineNo}: invalid type ${JSON.stringify(type)} (expected one of ${JSON.stringify(VALID_TYPES)})`,
    );
  }

  const symbol = get("asset_symbol");
  if (!symbol) {
    errors.push(`line ${lineNo}: asset_symbol is required`);
  }

  const currency = get("currency").toUpperCase();
  if (currency.length !== 3 || !/^[A-Z]{3}$/.test(currency)) {
    errors.push(`line ${lineNo}: invalid currency ${JSON.stringify(currency)} (expected a 3-letter code)`);
  }

  const quantityRaw = get("quantity");
  const quantity = Number(quantityRaw);
  if (quantityRaw === "" || Number.isNaN(quantity)) {
    errors.push(`line ${lineNo}: quantity must be numeric`);
  }

  const priceRaw = get("price");
  const price = Number(priceRaw);
  if (priceRaw === "" || Number.isNaN(price)) {
    errors.push(`line ${lineNo}: price must be numeric`);
  }

  const feesRaw = get("fees");
  let fees = 0;
  if (feesRaw !== "") {
    fees = Number(feesRaw);
    if (Number.isNaN(fees)) {
      errors.push(`line ${lineNo}: fees must be numeric`);
    }
  }

  const source = get("source");
  if (!source) {
    errors.push(`line ${lineNo}: source is required`);
  }

  if (errors.length > 0) {
    return { normalized: null, errors };
  }

  return {
    normalized: {
      date: dateStr,
      assetSymbol: symbol,
      assetName: get("asset_name") || null,
      assetClass,
      type,
      quantity,
      price,
      fees,
      currency,
      source,
      notes: get("notes") || null,
    },
    errors: [],
  };
}

/** Deterministic id so re-importing the same (or overlapping) CSV is a no-op. */
function transactionId(row: NormalizedRow): string {
  const key = [
    row.date,
    row.assetSymbol,
    row.assetClass,
    row.type,
    String(row.quantity),
    String(row.price),
    String(row.fees),
    row.currency,
    row.source,
  ].join("|");
  return createHash("sha256").update(key, "utf8").digest("hex");
}

async function getOrCreateAsset(
  client: PoolClient,
  symbol: string,
  name: string | null,
  assetClass: AssetClass,
  currency: string,
): Promise<{ id: number; created: boolean }> {
  const existing = await client.query<{ id: number }>("SELECT id FROM assets WHERE symbol = $1", [symbol]);
  if (existing.rows.length > 0) {
    return { id: existing.rows[0].id, created: false };
  }
  if (!name) {
    throw new CsvValidationError(
      `asset ${JSON.stringify(symbol)} does not exist yet — asset_name is required the first time a new symbol appears`,
    );
  }
  const inserted = await client.query<{ id: number }>(
    "INSERT INTO assets (symbol, name, asset_class, currency) VALUES ($1, $2, $3, $4) RETURNING id",
    [symbol, name, assetClass, currency],
  );
  return { id: inserted.rows[0].id, created: true };
}

/**
 * Validate and import a transactions CSV. All-or-nothing: if any row fails
 * validation, nothing is written. `client` must be a checked-out PoolClient
 * (not the Pool itself) so BEGIN/COMMIT/ROLLBACK apply to a single connection.
 */
export async function importTransactionsCsv(csvText: string, client: PoolClient): Promise<ImportStats> {
  const rawRecords = parseRows(csvText);

  const normalizedRows: NormalizedRow[] = [];
  const allErrors: string[] = [];
  rawRecords.forEach((raw, i) => {
    const lineNo = i + 2; // line 1 is the header
    const { normalized, errors } = validateAndNormalizeRow(raw, lineNo);
    if (errors.length > 0) {
      allErrors.push(...errors);
    } else if (normalized) {
      normalizedRows.push(normalized);
    }
  });

  if (allErrors.length > 0) {
    throw new CsvValidationError(`CSV validation failed:\n${allErrors.join("\n")}`);
  }

  const stats: ImportStats = { rowsRead: rawRecords.length, assetsCreated: 0, inserted: 0, skippedDuplicates: 0 };

  await client.query("BEGIN");
  try {
    for (const row of normalizedRows) {
      const { id: assetId, created } = await getOrCreateAsset(
        client,
        row.assetSymbol,
        row.assetName,
        row.assetClass,
        row.currency,
      );
      if (created) stats.assetsCreated += 1;

      const txnId = transactionId(row);
      const existing = await client.query("SELECT 1 FROM transactions WHERE id = $1", [txnId]);
      if (existing.rows.length > 0) {
        stats.skippedDuplicates += 1;
        continue;
      }

      await client.query(
        `INSERT INTO transactions
           (id, date, asset_id, asset_class, source, type, quantity, price, fees, currency, notes)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)`,
        [
          txnId,
          row.date,
          assetId,
          row.assetClass,
          row.source,
          row.type,
          row.quantity,
          row.price,
          row.fees,
          row.currency,
          row.notes,
        ],
      );
      stats.inserted += 1;
    }
    await client.query("COMMIT");
  } catch (err) {
    await client.query("ROLLBACK");
    throw err;
  }

  return stats;
}

File: portfolio/importer.py
"""Manual CSV import for the `transactions` table.

This is the universal fallback import path (spec section 1's "design rule"):
every live connector added later feeds into this same canonical schema, so a
broken connector never blocks you — you can always fall back to a CSV.

See sample_data/transactions_template.csv for the column layout and
sample_data/transactions_sample.csv for a filled-in (dummy) example.
"""

import csv
import hashlib
import sqlite3
from datetime import datetime
from pathlib import Path

REQUIRED_COLUMNS = {
    "date", "asset_symbol", "asset_class", "type",
    "quantity", "price", "fees", "currency", "source",
}
OPTIONAL_COLUMNS = {"asset_name", "notes"}
VALID_ASSET_CLASSES = {"stock", "etf", "cash", "gold", "crypto"}
VALID_TYPES = {"buy", "sell", "deposit", "withdrawal", "interest", "dividend"}


class CsvValidationError(Exception):
    """Raised when the CSV is malformed. Nothing is written to the DB when this is raised."""


def _read_rows(csv_path: str) -> list[dict]:
    with open(csv_path, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        header = set(reader.fieldnames or [])
        missing = REQUIRED_COLUMNS - header
        if missing:
            raise CsvValidationError(
                f"CSV is missing required column(s): {sorted(missing)}"
            )
        return list(reader)


def _validate_and_normalize_row(row: dict, line_no: int) -> tuple[dict, list[str]]:
    errors: list[str] = []

    def get(field: str) -> str:
        return (row.get(field) or "").strip()

    date_str = get("date")
    try:
        datetime.strptime(date_str, "%Y-%m-%d")
    except ValueError:
        errors.append(f"line {line_no}: invalid date {date_str!r} (expected YYYY-MM-DD)")

    asset_class = get("asset_class").lower()
    if asset_class not in VALID_ASSET_CLASSES:
        errors.append(
            f"line {line_no}: invalid asset_class {asset_class!r} "
            f"(expected one of {sorted(VALID_ASSET_CLASSES)})"
        )

    txn_type = get("type").lower()
    if txn_type not in VALID_TYPES:
        errors.append(
            f"line {line_no}: invalid type {txn_type!r} (expected one of {sorted(VALID_TYPES)})"
        )

    symbol = get("asset_symbol")
    if not symbol:
        errors.append(f"line {line_no}: asset_symbol is required")

    currency = get("currency").upper()
    if len(currency) != 3 or not currency.isalpha():
        errors.append(f"line {line_no}: invalid currency {currency!r} (expected a 3-letter code)")

    quantity = None
    try:
        quantity = float(get("quantity"))
    except ValueError:
        errors.append(f"line {line_no}: quantity must be numeric")

    price = None
    try:
        price = float(get("price"))
    except ValueError:
        errors.append(f"line {line_no}: price must be numeric")

    fees_raw = get("fees")
    fees = 0.0
    try:
        fees = float(fees_raw) if fees_raw else 0.0
    except ValueError:
        errors.append(f"line {line_no}: fees must be numeric")

    source = get("source")
    if not source:
        errors.append(f"line {line_no}: source is required")

    normalized = {
        "date": date_str,
        "asset_symbol": symbol,
        "asset_name": get("asset_name"),
        "asset_class": asset_class,
        "type": txn_type,
        "quantity": quantity,
        "price": price,
        "fees": fees,
        "currency": currency,
        "source": source,
        "notes": get("notes") or None,
    }
    return normalized, errors


def _transaction_id(row: dict) -> str:
    """Deterministic id so re-importing the same (or overlapping) CSV is a no-op."""
    key = "|".join([
        row["date"], row["asset_symbol"], row["asset_class"], row["type"],
        f"{row['quantity']:.10g}", f"{row['price']:.10g}", f"{row['fees']:.10g}",
        row["currency"], row["source"],
    ])
    return hashlib.sha256(key.encode("utf-8")).hexdigest()


def _get_or_create_asset(
    conn: sqlite3.Connection, symbol: str, name: str, asset_class: str, currency: str
) -> tuple[int, bool]:
    row = conn.execute("SELECT id FROM assets WHERE symbol = ?", (symbol,)).fetchone()
    if row:
        return row["id"], False
    if not name:
        raise CsvValidationError(
            f"asset {symbol!r} does not exist yet — asset_name is required the first "
            f"time a new symbol appears"
        )
    cur = conn.execute(
        "INSERT INTO assets (symbol, name, asset_class, currency) VALUES (?, ?, ?, ?)",
        (symbol, name, asset_class, currency),
    )
    return cur.lastrowid, True


def import_csv(csv_path: str | Path, conn: sqlite3.Connection) -> dict:
    """Validate and import a transactions CSV. All-or-nothing: if any row fails
    validation, nothing is written. Returns a stats dict."""
    rows = _read_rows(str(csv_path))

    normalized_rows = []
    all_errors = []
    for i, raw in enumerate(rows, start=2):  # line 1 is the header
        normalized, errors = _validate_and_normalize_row(raw, i)
        if errors:
            all_errors.extend(errors)
        else:
            normalized_rows.append(normalized)

    if all_errors:
        raise CsvValidationError("CSV validation failed:\n" + "\n".join(all_errors))

    stats = {"rows_read": len(rows), "assets_created": 0, "inserted": 0, "skipped_duplicates": 0}
    try:
        for row in normalized_rows:
            asset_id, created = _get_or_create_asset(
                conn, row["asset_symbol"], row["asset_name"], row["asset_class"], row["currency"]
            )
            if created:
                stats["assets_created"] += 1

            txn_id = _transaction_id(row)
            exists = conn.execute("SELECT 1 FROM transactions WHERE id = ?", (txn_id,)).fetchone()
            if exists:
                stats["skipped_duplicates"] += 1
                continue

            conn.execute(
                """INSERT INTO transactions
                   (id, date, asset_id, asset_class, source, type, quantity, price, fees, currency, notes)
                   VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
                (
                    txn_id, row["date"], asset_id, row["asset_class"], row["source"], row["type"],
                    row["quantity"], row["price"], row["fees"], row["currency"], row["notes"],
                ),
            )
            stats["inserted"] += 1
        conn.commit()
    except Exception:
        conn.rollback()
        raise

    return stats
